OpenRA.Mods.Common/Activities/Move/AttackMoveActivity.cs

```diff
@@ -20,92 +20,80 @@ namespace OpenRA.Mods.Common.Activities
 	public class AttackMoveActivity : Activity
 	{
 		readonly Func<Activity> getInner;
-		public readonly bool IsAssaultMove;
-		Activity inner;
-		Activity attack;
+		readonly bool isAssaultMove;
 		AutoTarget autoTarget;
-		bool moving;
+		ConditionManager conditionManager;
+		AttackMove attackMove;
+		int token = ConditionManager.InvalidConditionToken;
 
 		public AttackMoveActivity(Actor self, Func<Activity> getInner, bool assaultMoving = false)
 		{
 			this.getInner = getInner;
 			autoTarget = self.TraitOrDefault<AutoTarget>();
-			moving = false;
-			IsAssaultMove = assaultMoving;
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+			attackMove = self.TraitOrDefault<AttackMove>();
+			isAssaultMove = assaultMoving;
 		}
 
-		public override Activity Tick(Actor self)
+		protected override void OnFirstRun(Actor self)
 		{
-			if (IsCanceling)
-			{
-				if (attack != null)
-				{
-					attack = ActivityUtils.RunActivity(self, attack);
-					return this;
-				}
+			// Start moving.
+			QueueChild(self, getInner());
 
-				if (inner != null)
-				{
-					inner = ActivityUtils.RunActivity(self, inner);
-					return this;
-				}
+			if (conditionManager == null || attackMove == null)
+				return;
 
-				return NextActivity;
-			}
+			if (!isAssaultMove && !string.IsNullOrEmpty(attackMove.Info.AttackMoveScanCondition))
+				token = conditionManager.GrantCondition(self, attackMove.Info.AttackMoveScanCondition);
+			else if (!string.IsNullOrEmpty(attackMove.Info.AssaultMoveScanCondition))
```

Comment:
This also needs to check `isAssaultMove` to prevent it granting the assault move condition for everything when `AttackMoveScanCondition` isn't defined (which is the case for the default mods).

Reply:
The regression can be seen by ordering an attack-move past an enemy structure. Expected behaviour is for the units to ignore the structure during a regular attack move, but attack for an assault move.